### ISSUE TITLE
[feature] define remote- and upload-urls in conanfile

### ISSUE THREAD

**Issue by stefaneicher** [conan-io/conan#8322]
Could you add something in the conan file such that a conan file self contained could define where he want o to go search for dependencies and where it is intended to upload the package to?

Similar to https://maven.apache.org/pom.html#distribution-management

What if the conan file could look like this:

from conans import ConanFile, CMake

class MypackageConan(ConanFile):
    name = "mypackage"
    version = "1.0"
    ...
    # Define where this conanfile should go search for dependencies
    # This has prio over the "conan remote list" repos
    # Think of a list like remotes and appended (conan remote list)
    dependency_remotes = "https://api.bintray.com/conan/disroop/conan"
    
    # repo where to upload the package
    # or for example implement a upload() method to implement a if else logic depending weather it is a development or release pacakge
    upload_remotes= "https://api.bintray.com/conan/disroop/conan"

currently, I have to orchestrate this topic outside of the conan file.  I have seen that https://github.com/conan-io/conan-package-tools adresses the issue. But wi would like to have this feature a first class citicen added in the conan file.
My proposal would help to make such thing easier to setup and communicate.
Further, I would not need to change the shard state by changing the "conan remotes". This would also give me confidence that parallel builds would not act upon an unpredictable "conan remote list" -lists. 

What do you think about my suggestion? What facts did I miss in my understanding?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @stefaneicher 

What you are pointing there about Maven is very relevant, as it is a well known issue for many devops good practices. JFrog has tons of experience in this, and one of the things that is necessary many times in Maven server repositories is to patch the Maven files to remove the remote information. Otherwise you cannot cache them in your own repos, you cannot do binary promotions in your pipelines, you cannot replicate binaries among distributed servers for different teams around the world, etc.

So in Conan this is a meditated decision, to separate the definition of "what" from the definition of "where." A given recipe can be uploaded and belong to different remotes, and its requirements and dependencies can be found in different remotes as well. Hardcoding the repositories in the recipes goes against these practices. Given them more or less priority (appending to the existing ones), doesn't fix the issue, because there will always be cases that mandate the opposite priority. That complexity shouldn't be in the client side, in the same way it is not hardcoded in the source code of a git project. You can clone a repo at anytime, push it to any other fork without problem and without changing a single line of code.

Using virtual repositories in the server that aggregates different remotes is possibly one of the recommended ways of making your flows simpler and more robust. 

I hope this clarify a bit this design decision, please let me know. Thanks!

**Comment by stefaneicher** [comment]
Hi @memsharded ,

Thanks for your answer. Good to see that this was a mediated decision. It containes the missing explanation for me. 
I have 10 years experience with maven and never had such problems. But your perspective is obviously broader.
And yes we are using virtual repos in artifactory.

P.s. I also miss a bit the comfot of the maven multi modules alt least wit the intellij maven integration. We also use the workspaces. But I will layout my question in a next issue I will create.

We created this https://github.com/disroop/vs-code-conan. And it needs improvements.
We are designing a solution similar to the maven integration in intellij. Would be nice to get your opinion before we start to spend many hours if you could spot conceptual errors right at the idea or design level.

**Comment by memsharded** [comment]
> I have 10 years experience with maven and never had such problems. But your perspective is obviously broader.
And yes we are using virtual repos in artifactory.

Yes, it might be possible that it can work in some scenarios without problems. But it can also be very problematic in other contexts. Virtual repositories work for Maven because Artifactory does a lot of efforts to make it happen, it can patch the files on the fly to account for the removal of this to make virtual repos work. And it is relatively doable because it is xml, but still problematic because the checksums will change, but Artifactory knows how to workaround that. Now Conan recipe is python, so changing it there would be more difficult. And besides the checksum changing, in Conan that would mean that the revisions would be broken, because they are based on the contents checksum.

We will likely discuss how to improve reproducibility of the things that are not stored in a lockfile (which would be the main mechanism for reproducibility), like the remotes somewhere. It might be in the lockfiles themselves, but also lockfiles might be used to work with multiple, different remotes, so it might be something that is stored together with the lockfiles. It is likely that the CI/CD training that we are preparing in the JFrog Academy (https://academy.jfrog.com/path/conan) might contain something in this line.

Thanks for telling about the https://github.com/disroop/vs-code-conan integration. For your info, we are aware that the workspaces is not a very mature feature, it is limited, and it really needs some work, but it didn't get enough priority yet. We are trying to improve it from the roots, revisiting the concept of "editable" packages, and exploring the definition of "layouts" of packages in local edition mode. Hopefully we can keep improving this feature from there. 

Yes, please let us know about your current plans and view for https://github.com/disroop/vs-code-conan. If you want to explain it in a dedicated issue, great, but if you need to talk something privately first, let us know in info@conan.io, maybe we can find another way to sync. Thanks!

**Comment by stefaneicher** [comment]
Hi @memsharded 

Thanks for your answer. Sorry for my late answer. We (@mymichu and me) just did not find time to continue our work on the plugin as intended. But we definitely would like to followup with your offer to sync and will contact you in den next weeks.

Regards Stefan

**Comment by memsharded** [comment]
We have recently added an entry in the FAQ (https://docs.conan.io/en/latest/faq/using.html#can-conan-use-git-repositories-as-package-servers) that comments about not having remotes or any other "absolute" package origin reference in the conanfile recipes. 

So I think this can be closed.

There might be still the case for a local definition of remotes, besides the cache remote configuration. At the moment is straightforward to run a ``conan config install <local repo remotes.txt file>`` to install that file in the cache, so a repository can contains its own remotes definition to be used by users for convenience. We could further automate it and do kind of an automatic detection of the file and doing an addition/merge to the current existing repos (on the other hand, seems a relatively risky feature if that is applied from repos you clone from elsewhere). In any case, this would be another feature to be discussed in a separate issue. Thanks!
